app/javascript/components/upload/ExpressionFileForm.js

```diff
@@ -5,54 +5,67 @@ import Select from 'lib/InstrumentedSelect'
 import MTXBundledFilesForm from './MTXBundledFilesForm'
 import FileUploadControl, { FileTypeExtensions } from './FileUploadControl'
 import { TextFormField, SavingOverlay, SaveDeleteButtons } from './form-components'
+import { findBundleChildren, validateFile } from './upload-utils'
+
+const REQUIRED_FIELDS = [{ label: 'species', propertyName: 'taxon_id' },
+  { label: 'Biosample input type', propertyName: 'expression_file_info.biosample_input_type' },
+  { label: 'Library preparation protocol', propertyName: 'expression_file_info.library_preparation_protocol' },
+  { label: 'Modality', propertyName: 'expression_file_info.modality' }]
+const RAW_COUNTS_REQUIRED_FIELDS = REQUIRED_FIELDS.concat([{
+  label: 'units', propertyName: 'expression_file_info.units'
+}])
 
 /** renders a form for editing/uploading an expression file (raw or processed) and any bundle children */
 export default function ExpressionFileForm({
   file,
+  allFiles,
   updateFile,
   saveFile,
   deleteFile,
   addNewFile,
-  handleSaveResponse,
   fileMenuOptions,
-  rawCountsOptions=[],
-  associatedChildren
+  rawCountsOptions
 }) {
-
+  const associatedChildren = findBundleChildren(file, allFiles)
   const speciesOptions = fileMenuOptions.species.map(spec => ({ label: spec.common_name, value: spec.id }))
   const selectedSpecies = speciesOptions.find(opt => opt.value === file.taxon_id)
   const isMtxFile = file.file_type === 'MM Coordinate Matrix'
   const isRawCountsFile = file.expression_file_info.is_raw_counts
+
+  const allowedFileTypes = isMtxFile ? FileTypeExtensions.mtx : FileTypeExtensions.plainText
+  const requiredFields = isRawCountsFile ? RAW_COUNTS_REQUIRED_FIELDS : REQUIRED_FIELDS
+  const validationMessages = validateFile({ file, allFiles, allowedFileTypes, requiredFields })
+
   const associatedRawCounts = file.expression_file_info.raw_counts_associations.map(id => ({
     label: rawCountsOptions.find(rf => rf.value == id)?.label,
     value: id
   }))
 
   return <div className="row top-margin" key={file._id}>
     <div className="col-md-12">
-      <form id={`clusterForm-${file._id}`}
+      <form id={`expForm-${file._id}`}
```

Comment:
Let's use hyphen-case for attribute values, per convention:

```suggestion
      <form id={`exp-form-${file._id}`}
```